Add more examples. (#196)
WICG/sanitizer-api: explainer.md

All of these would have had identical results if the "unsafe" variants had
been used.

> ### Parsing XML

Reviewer: Should this say "XML documents" instead, since `setHTML` parses as HTML?

  attributes: [ "class" ]  // Allows only class attributes.
      // Output with "safe" and "unsafe" methods are the same for this config.
};
const config_disallow_style_definitions = {
  removeElements: [ "style" ],  // Allows the defaults, but without <style>.
  removeAttributes: [ "class", "style" ]  // No style or class attribute either.
      // And not XSS-y stuff, either, if used with a "safe" method.
      // Output with "safe" and "unsafe" methods might be quite different.
};
```

You may also wish to remove elements, but retain their contents. This is
chiefly useful to remove unwanted formatting from user input, while
preserving its textual content.

```js
const config_that_removes_elements_but_preserves_their_children = {
 flattenElements: ["span", "em", "u", "s", "i", "b"]
};

element.setHTML(
  "Fancy <b>text</b> with <span style='color:blue'>pizzazz</span>.",
  { sanitizer: config_that_removes_elements_but_preserves_their_children });
  // <div>Fancy text with pizzazz.</div>
```

There is no `flattenAttribute` because attribute nodes do not have children.

Flatten applies to its immediate children, i.e. to one level. Combining

> `allowElements` with `flattenElements` lets you keep some formatting, but all
> the text content:
> 
> ```js
> const config_flatten_spans = {
>   allowElements: ["b", "i"],

Reviewer 2: ```suggestion
`elements` with `flattenElements` lets you keep some formatting, but all
the text content:

```js
const config_flatten_spans = {
  elements: ["b", "i"],
```

};
```

> [!NOTE]
> The `config_with_namespaces` example contains multiple entries for the same
> element or attribute, to illustrate the syntax. Note that this isn't actually
> allowed.

### Configuration Options: Allowing or removing (blocking) or flattening

There are two ways you can build up a config: Specify the elements & attributes
you wish to allow. This is easy to read and makes it easy to understand what
to expect in the sanitizer output. Or you can specify what elements & attributes
you wish to remove. Or to block, as other sanitizer libraries might call it.
This effectively specifies the sanitizer output relative to the built-in list.
This can be useful if you wish to mostly retain the built-in defaults.

```js
const config_allow_some_formatting = {
  elements: [ "div", "p", "em", "b", "img" ],  // Allows only 5 elements.
  attributes: [ "class" ]  // Allows only class attributes.
      // Output with "safe" and "unsafe" methods are the same for this config.
};
const config_anything_but_style_definitions = {
  removeElements: [ "style" ],  // Allows a lot of things. But not <style>.

>   removeAttributes: [ "class", "style" ]  // No style or class attribute either.

Reviewer 3: Nit: maybe instead of "anything" or "a lot of things", let's say "the defaults except ..."

Author: Done.

      // And not XSS-y stuff, either, if used with a "safe" method.
      // Output with "safe" and "unsafe" methods might be quite different.
};
```

You may also wish to remove elements, but retain their contents. This is
chiefly useful to remove unwanted formatting from user input, while
preserving its textual content.

```js
const config_that_removes_elements_but_preserves_their_text_content = {
 flattenElements: ["span", "em", "u", "s", "i", "b"]
};

element.setHTML(

>   "Fancy <b>text</b> with <span style='color:blue'>pizzazz</span>.",

Reviewer 4: Would adding an example that flattens only one level help here?
e.g. 
```
const config_flatten_spans = {flattenElements: ["span"]};
element.setHTML("Fancy <span style='color:blue'> text with <b>pizzazz</b></span>.", { sanitizer: config_flatten_spans});
  // <div>Fancy text with <b>pizzazz</b>.</div>
```

element.setHTML(example_with_shadow_root);
//<div>
//  #shadow-root: <b>In the shadows.</b>

element.setHTMLUnsafe(example_with_shadow_root);
//<div>
//  #shadow-root: <b onlick='alert(1)'>In the shadows.</b>

element.setHTML(example_with_shadow_root, {sanitizer: {allowShadowRoot: false}});
// <div>

element.setHTMLUnsafe(example_with_shadow_root, {sanitizer: {allowShadowRoot: false}});
// <div>
```

### Configuration Errors

The configuration allows expressing redundant or even contradictory options.
For example, allowing and removing the same element. In cases where the
meaning of a configuration dictionary isn't clear, we will
throw a `TypeError` instead of making a best effort attempt at interpreting
the configuration. A well-formed configuration has the following properties:

* It contains either an allow-list or a remove-list, but not both.
  * This applies to both element and attribute lists, seperately.
  * Note that any config with both, an allow-list and remove-list, can be
    rewritten by removing the remove-list items from the allow-list and then
    droping the remove-list entirely.

>   * A config with an allow-list and a flatten-list makes sense since the items

Reviewer 4: What about flatten-list + block-list? I think it also makes sense.

1. When passing an allow-list, the implicit default is to block, so passing a block-list doesn't make sense.
2. When passing a block-list, the implicit default is to allow.
3. When passing both, the implicit default is to block, but then it doesn't make sense to define the block-list (back to point 1).

In all cases, passing a flatten-list makes sense because it's changing the implicit default.

    in the flatten list preserve their child contents, while the allow-list does

>     not.

Author: :-)  I can no longer figure out what I was actually trying to say there... I removed that sentence in favour of "Both allow-lists and remove-lists can be combined with flatten-lists."

> * Any allow-, remove-, or flatten-list should contain each name at most once.

Reviewer 4: Each name must not appear in a given list (allow-, flatten-, remove-) several times, but the same name shouldn't appear in several list either. Is this missing?

Author: True. I adapted the description.

  * This would apply to short forms as well.
    E.g., `["div", { name: "div", namespace: "http://www.w3.org/1999/xhtml" }]`
    contains the same name twice and would thus throw.
  * While lists with duplicate element or attribute names could be coalesced,
    it is ambiguous what the meaning of duplicate elements with different
    element-dependent attribute lists would be.

> * The name must be set. If an `attributes` or `removeAttributes` key is present,

Reviewer 3: Nit: s/name/element name/

Author: This was meant as applying to both element and attribute lists, so that a dictionary without a name-key set would be invalid.

Reviewer 3: Fair enough. Keep it :)

>   it must be non-empty. It may contain an empty list, though.

Author: Yes. Removed it.

```js
// Mixing allow and block lists throws.
const config_that_mixes_allow_and_block_lists = {
    elements: ["i", "u"],
    removeElements: ["u", "s"],
};
element.setHTML("bla", {sanitizer: config_that_mixes_allow_and_block_lists}); // throws

// Mixing allow and flatten lists works.
const config_that_retains_simple_styling_but_most_text = {
  elements: ["p", "b", "i"],
  flattenElements: ["div", "span", "em", "u", "s", "li"],
};
const styled_text = "<p>Some <span style='color: blue'>colourful</span> <u>styled</u> <b>text</b>";

// <div><p>Some colourful styled <b>text</b></p></div>
element.setHTML(styled_text, {sanitizer: config_that_retains_simple_styling_but_most_text});

// Duplicate entries throw.
const config_with_dupes = {
  elements: [ "div", { name: "div", namespace: "http://www.w3.org/1999/xhtml" } ]
};
element.setHTML("bla", {sanitizer: config_with_dupes});  // throws.

const config_with_dupes2 = {
  elements: [
    { name: "div", attributes: ["class"] },
    { name: "div", attributes: ["style"] }
  ] };
element.setHTML("bla", config_with_dupes2);  // throws.

> // Undefined attributes. What does it mean?

Reviewer 3: Is that a WebIDL questions?
I don't even know where to look that up... @annevk can you help?

Author: It was meant to highlight that `{ name: "div", attributes: undefined }` could mean the same as either `{ name: "div" }` (allows the default attrs for `div`) or `{ name: "div", attributes: [] }` (allows no attrs for `div`). At least to me it's not obvious what interpretation to pick.

But.. you're right. Maybe WebIDL solves that problem for us and interprets in a way that does make it clear. I didn't think of that.

Author: Thanks. Removed the example.

```js
const config_anything_but_style_definitions = {
  blockElements: ["style"],
  blockAttributes: ["style", "id"],
};
```

If you know your input will contain elaborate markup and you only wish to
retain some of it, but you want to retain the full textual content, you
can use `dropElement`. This will remove only elements but retain their children.
There is no `dropAttribute` because attribute nodes do not have children.

```js
const config_that_drops_things = {
  dropElements: ["span", "em", "u", "s", "i", "b"]
};
```

How do multiple element list types interact? In order for an element to
pass, it must pass _all_ lists.

You can use allow-lists and block-lists together, but note that this doesn't
usually make much sense since any combination of allow- and block-lists can
be rewritten as an allow-lists of equal or shorter length by removing any
block-list element from the allow-list, or dropping it altogether if there
was no matching allow-list entry in the first place.

```js
// equivalent to:  { allowElements: ["i"] }
const config_that_mixes_allow_and_block_lists = {
    allowElements: ["i", "u"],
    blockElements: ["u", "s"],

> };

Reviewer 3: That's all correct and matches my understanding. I think it would make more sense for a reader if we defined and then pointed to the processing model (Roughly "check first drop / block first, then allow." like we already have in the "determine the sanitize action for an element" algorithm)

```

### Configuring attributes per element

A common use case is to allow or block all instances of a given attribute,
but this isn't always sufficient. One may also act on attributes on specific
elements.

In the example `config_text_with_only_simple_styling` in the previous chapter
we have allowed the `src` attribute on any of the eight allowed elements.
If one wanted to allow `class` everywhere, but `src` only on `<img>`, the
following would do:

```js
const config_with_element_specific_attributes = {
  allowElements: [
    "p", "b", "i",
    { name: "img", attributes: [ "src" ] }
  ],
  allowAttributes: ["class"],
};
```

If you want to drop `src` attributes from `<input>` elements but retain them
elsewhere, you can use:

```js
const drop_src_attribute_from_input = {
  dropElements: [{ name: "input", attributes: ["src"]}],
}
```

TODO: What does `{ name: "bla", attributes: [] }` mean? My gut feeling would
  be that `attributes: []` is a no-op and it allows `<bla>` plus whatever
  attributes are allowed globally. But if so, the only way to describe an
  element without any attributes in presence of a global `allowAttributes`

>   would be to add a dropElements for it.

Reviewer 3: To be clear, this question is for `allowAttributes: { name: "bla", attributes: [] }` right?
Hm, what if empty actually meant "no attributes for this element"? That would make the list look-up "just work" and allow special-casing elements where the global `allowAttributes` shouldn't apply (I am assuming both rules must allow the attribute).

Author: I meant `allowElements`, not `allowAttributes`.

Given the input `"<img src=bla srcset=blubb alt=blerk>"`, what do these do:

* `allowElements: [{ name: "img", attributes: [] }]`
* `allowElements: ["img"]`
* `allowElements: `[{name: "img", attributes: undefined }]`

My gut feeling is that these three should mean the same thing. (I could certainly be convinced otherwise.)

On: _"allow special-casing elements where the global `allowAttributes` shouldn't apply (I am assuming both rules must allow the attribute)."_

* Would `allowElements: [{ name: "img", attributes: [] }], allowAttributes: ["src", "srcset"]` turn the input example above into "<img>", because the attributes are not allowed by both rules?

I'm worried that we're making the "grab bag" of elements and attributes very laborious, which was the primary use case we found when looking at existing sanitizer usage.

Reviewer 3: 1. Yes, I think all of those three examples should resulti n `<img>` with no attributes.
2. Great point. We shouldn't make this too cumbersome, I agree. I guess this could work if our step for "finding the allowed attributes for this current element" were to build the union of what `allowElement` and `allowAttributes` are specifying. But then, how would you build an instruction that is "_allow an element `img` with no attributes at all, not even the ones specified elsewhere_". Can't think of something but happy to go with a suggestion from you.

TODO: I find the syntax weird. In the allowElements case, the `img` entry
  matches (and hence allows) both the element and the attribute. In the drop
  case it matches only the attribute (but not the element). If we were to
  define this identically in both cases - always match element and attributes -
  then it simply wouldn't work for drop, because a drop-rule with attributes

>   would automatically drop the element (and hence all attributes with it).

Reviewer 3: Can you expand? I'm not sure I fully understand the example again. 
My understanding:
- We go through all elements and if it is allowed, the list of attributes
- We go through all attributes and keep them if allowed

These "keep if allowed" steps would go through drop/block/allow for elements and block/allow for attributes in that order. If all of those 3 (for elements) or 2 (for attributes) lists allow the node, then it should be kept.

If the attribute-list is a list of attributes, we would match as "the attribute is in list" just according to the attribute. If the attribute-list is a list of attributes+elements, then it would only match as "is in list" if both element & attribute fit.

Does that make sense? Did I misunderstand the example?

Author: What does `blockElements: [{ name: "img", attributes: ["srcset"] }]` mean? Does it mean - given the sample input from the earlier comment - an empty tree, or `<img src=bla alt=blerk>`? If the former, why have the `attributes:` key at all? If the latter, then what does `blockElements: [{ name: "img", attributes: [] }]` mean?

The currently proposed syntax allows multiple clauses for one element. What do these mean:

* `allowElements: [{ name: "img", attributes: ["src"]}, { name: "img", attributes: ["srcset"]}]`
* `blockElements: [{ name: "img", attributes: ["src"]}, { name: "img", attributes: ["srcset"]}]`

(Btw, I'm not complaining; I'm just trying to sound out the edge cases.)

Reviewer 3: From the top of my head (still not fully back at my desk mentally), I'm not sure what the attribute-pieces even means for the `blockElements` case. Block the element if it has those attributes? Block the attributes but only on these elements? I guess the latter..

How do multiple attribute rules interact? An attribute is allowed, if at
least one allow-rule (global or per element) allows it, and no drop

> rule drops it.

Reviewer 3: I think we would go through the lists in order. If it is dropped, then we don't continue looking into the allow list. Similar to existing spec text. No?

Reviewer 3: Discussion from today's meeting to explain the issue at hand a bit better.

Example: allowing `{name: "div", attribute: "foo"}, {name: "div", attribute: "bar"}`and input `<div foo="" bar="">`

If the algorithm would go through each entry individually, leading to `<div>`.
If the algorithm would collect all entries and combine it would return the input.

```js
const config_allow = {
  allowElements: [ "div", "p", "em", "b" ]  // Allows only those four elements.
  allowAttributes: [ "class" ]  // Allows only class attributes.
      // Output with "safe" and "unsafe" methods are the same for this config.
};
const config_block = {

>   blockElements: [ "style" ],  // Allows a lot of things. But not <style>.

Reviewer 3: s/Allows a lot of things/Allows the default, except <style>/, maybe?

  blockAttributes: [ "class", "style" ]  // No style or class attribute either.
      // And not XSS-y stuff, either, if used with a "safe" method.
      // Output with "safe" and "unsafe" methods might be quite different.
};
```

### Configuration List-Types: Allow-lists, Blocks-lists, and Drop-lists

For elements there are three types of lists: Allow-lists, Block-lists
and Drop-lists. These correspond to three related but somehow different use
cases:

If you have a specific set of functionality in mind that you wish to
retain and would like to drop everything else as a matter of caution, this
can be accomplished with allow lists. This is probably the most common case.

> For example, you might want allow user comments with simple styling, but

Reviewer 3: Nit: they are called "formatting elements"

also disallow new styles or `id`s not already declared in your document:

```js
const config_anything_but_style_definitions = {
  blockElements: ["style"],
  blockAttributes: ["style", "id"],
};
```

If you know your input will contain elaborate markup and you only wish to
retain some of it, but you want to retain the full textual content, you
can use `dropElement`. This will remove only elements but retain their children.
There is no `dropAttribute` because attribute nodes do not have children.

```js
const config_that_drops_things = {
  dropElements: ["span", "em", "u", "s", "i", "b"]
};
```

### Configuring attributes per element

A common use case is to allow or block all instances of a given attribute,
but this isn't always sufficient. One may also act on attributes on specific
elements.

In the example `config_text_with_only_simple_styling` in the previous chapter
we have allowed the `src` attribute on any of the eight allowed elements.
If one wanted to allow `class` everywhere, but `src` only on `<img>`, the
following would do:

```js
const config_with_element_specific_attributes = {
  allowElements: [
    "p", "b", "i",

>     { name: "img", attributes: [ "src" ] }

Reviewer 3: `allowElements` is just elements.

  ],
  allowAttributes: ["class"],
};
```

If you want to drop `src` attributes from `<input>` elements but retain them
elsewhere, you can use:

```js
const drop_src_attribute_from_input = {
  dropElements: [{ name: "input", attributes: ["src"]}],
}

> ```

Reviewer 3: Should this drop the element or the attribute?

Reviewer 3: Looking back at https://github.com/WICG/sanitizer-api/issues/181, we had `dropElements` just accepting elements and `dropAttributes` to accept `{name: 'src', elements: ['input'] }`, which is a bit inverse of what the allow-properties do

An element with no allowed attributes follows from these definitions. For an
element that allows any of the default-allowed attributes, you can use the
string `"*"`.

```js
const config_div_without_anything = {

>   allow_elements: [ { name: "div", attributes: [] } ]

Reviewer 3: nit: s/allow_elements/allowElements/

};
const config_div_with_everything = {
  allow_elements: [ { name: "div", attributes: "*" } ]
};
```

### Matching any attribute on a given element

> For some use cases, one may wish to allow (or drop) any attribute:

Reviewer 3: same question as above

The configuration allows expressing redundant or even contradictory options. In
cases where the meaning of a configuration dictionary isn't clear, we will
throw a `TypeError` instead of making a best effort attempt at interpreting
the configuration. A well-formed configuration has the following properties:

* It contains either an allow-list or a block-list, but not both.
  * This applies to both element and attribute lists, seperately.
  * A config with both, an allow-list and block-list, can be rewritten to
    remove the block-list items from the allow-list and then drop the allow-list
    entirely.
  * A config with an allow-list and a drop-list makes sense since the items
    in the drop-list preserve their child contents, while the allow-list does
    not.
* Any allow-, drop-, or block-list should contain each name at most once.
  * This would apply to short forms as well.
    E.g., `["div", { name: "div", namespace: "http://www.w3.org/1999/xhtml" }]`
    would throw.
  * While lists with duplicate element or attribute names could be coalesced,
    it is ambiguous what the meaning of duplicate elements with different
    element-dependent attribute lists would be.

> * The name must be set. If an `attributes` key is present, it must be non-empty.

Reviewer 3: non-empty? We have "empty array" examples above. Maybe disambiguate
